dali/pipeline/operators/arithmetic/arithmetic.h

```diff
@@ -0,0 +1,268 @@
+                     shape[sample_idx].num_elements()});
+    if (task_idx != prev_task_idx) {
+      ranges[task_idx].begin = descs.size() - 1;
+      ranges[task_idx].end = descs.size();
+    } else {
+      ranges[task_idx].end = descs.size();
+    }
+    prev_task_idx = task_idx;
+  }
+  return std::make_tuple(descs, ranges);
+}
+
+/**
+ * @brief Divide the shape into groups of linear extents, separated evenly between threads
+ */
+inline std::tuple<std::vector<TileDesc>, std::vector<TileRange>> GetTiledCover(
+    const kernels::TensorListShape<> &shape, int extent, int num_tasks) {
+  Index total_elements = shape.num_elements();
+  std::vector<TileDesc> descs;
+  std::vector<TileRange> ranges(num_tasks, {0, 0});
+  Index elements_per_thread = total_elements / num_tasks;
+  // One extent covers all
+  if (total_elements < extent || elements_per_thread == 0) {
+    return GetCover(shape, num_tasks);
+  }
+  int task_idx = 0;
+  int prev_task_idx = -1;
+  std::vector<Index> covered_by_thread(num_tasks, 0);
+  for (int sample_idx = 0; sample_idx < shape.num_samples(); sample_idx++) {
+    int extent_idx = 0;
+    for (Index covered = 0; covered < shape[sample_idx].num_elements(); covered += extent) {
+      auto actually_covered =
+          std::min(static_cast<Index>(extent), shape[sample_idx].num_elements() - covered);
+      descs.push_back({sample_idx, extent_idx, task_idx, actually_covered, extent});
+      // We either cover a full extent, of what is left in the sample
+      covered_by_thread[task_idx] += actually_covered;
+      // Update range information
+      if (task_idx != prev_task_idx) {
+        ranges[task_idx].begin = descs.size() - 1;
+        ranges[task_idx].end = descs.size();
+      } else {
+        ranges[task_idx].end = descs.size();
+      }
+      prev_task_idx = task_idx;
+      // If we covered what we need, proceed to next thread
+      if (covered_by_thread[task_idx] >= elements_per_thread && task_idx < num_tasks - 1) {
+        task_idx++;
+      }
+      extent_idx++;
+    }
+  }
+  return std::make_tuple(descs, ranges);
+}
+
+/**
+ * @brief Recurse over expression tree, fill the missing types of TensorInputs
+ *
+ * @param expr
+ * @param ws
+ * @return DALIDataType
+ */
+template <typename Backend>
+DLL_PUBLIC DALIDataType PropagateTypes(ExprNode &expr, const workspace_t<Backend> &ws) {
+  if (expr.GetNodeType() == NodeType::Constant) {
+    return expr.GetTypeId();
+  }
+  if (expr.GetNodeType() == NodeType::Tensor) {
+    auto &e = dynamic_cast<ExprTensor&>(expr);
+    expr.SetTypeId(ws.template InputRef<Backend>(e.GetMappedInput()).type().id());
+    return expr.GetTypeId();
+  }
+  if (expr.GetSubexpressionCount() == 2) {
+    auto left_type = PropagateTypes<Backend>(expr[0], ws);
+    auto right_type = PropagateTypes<Backend>(expr[1], ws);
+    expr.SetTypeId(TypePromotion(left_type, right_type));
+    return expr.GetTypeId();
+  }
+  DALI_FAIL("Only binary expressions are supported");
```

Comment:
You could add info here: "got expression with N elements"

Reply:
Done